### ISSUE TITLE
[BUG] Failed to load sound driver on Latitude 7320 Detachable 

### ISSUE THREAD

**Issue by prof-dr-pepper** [thesofproject/linux#3161]
**Describe the bug**
Sound driver seems to fail to load properly on Dell Latitude 7320 Detachable (see hardware probe here https://linux-hardware.org/?probe=21c50317b2) and therefore sound is not available at the device on linux.
Even though the sound adapter (hardware ID: 8086:a0c8) seems to be supported by the linux kernel the device does not work. Browsing on the internet this issue seems to happen on all Dell Latitude 7320 Detachable systems. I tested with kubuntu 21.04 with kernel 5.11.0-34-generic (however same issue occurs with latest Manjaro linux and kernels up to 5.14). Furthermore looking at dmesg I saw a lot of (thousands and millions) suppressed IRQ callbacks that keep the CPU busy (one core seems to be busy for about 50% because of that). Not loading the driver (via grub: snd_hda_intel.dmic_detect=0) the CPU issue is gone but of course then sound will be disabled.

**To Reproduce**
1. Run kubuntu 21.04 live image (sha256sum of ISO: e705e84ed8fa820f1b7f4d286d24e0eb97561a5b2d47421bf7ea0381103788b7) from USB stick
2. After boot look at dmesg (my dmesg: https://github.com/thesofproject/sof/files/7172540/dmesg.txt)

**Reproduction Rate**
Issue is reproducible all the time (also seems with other linux kernel versions and linux distros).

**Expected behavior**
Linux boots up and loads driver without issues. Sound is working properly and CPU usage is near 0.

**Impact**
Using linux without sound is annoying at the least but during home office makes it also useless during audio conferences (however video is also not working on the device but that's another issue because there exists not webcam driver for this device yet). ;-)

**Environment**
- Dell Latitude 7320 Detachable
- Kubuntu 21.04 64bit (see ISO hash above)
- Stock ubuntu linux kernel 5.11.0-34-generic
- HW probe: https://linux-hardware.org/?probe=21c50317b2
- Docking station with SSD for linux installation was attached during HW probe (however issue also occurs when booting from kubuntu 21.04 live USB stick)

**Screenshots or console output**
Here you can find my dmesg output and alsa-info directly after boot:
[dmesg.txt](https://github.com/thesofproject/sof/files/7172540/dmesg.txt)
http://alsa-project.org/db/?f=e7ce23315aee536587174f7f58f79812c1166d1a

Please let me know if you need more information and how I can provide it.

Thanx,
Marcus

**Comment by plbossart** [comment]
Thanks @prof-dr-pepper. I can see this: System SKU:        0A45, which is unfortunately a device we haven't seen before. I'll contact Dell and ask for details.

**Comment by prof-dr-pepper** [comment]
Let me know if there is anything I can help with. I am software developer myself (not a kernel developer, but at least fluent in C/C++). ;-)

**Comment by plbossart** [comment]
> Let me know if there is anything I can help with. I am software developer myself (not a kernel developer, but at least fluent in C/C++). ;-)

I will take you up on your offer, given that I don't have access to the hardware, we have to rely on the community for feedback. The main ask is to be able to recompile your own kernel, using our latest development Git tree. we may also try with linux-stable and patches. Thanks!

**Comment by prof-dr-pepper** [comment]
Sure. So I'll try this kernel https://github.com/thesofproject/linux but which branch? Normally I would assume "master" but in this case it looks more like "topic/sof-dev". So "topic/sof-dev"?

**Comment by plbossart** [comment]
yes. topic/sof-dev, with possibly a PR test before we merge to topic/sof-dev. we also generate our own configs with https://github.com/thesofproject/kconfig, specifically kconfig-sof-default.sh to only compile the minimum amount of code for audio tests.

**Comment by prof-dr-pepper** [comment]
> with possibly a PR test before we merge to topic/sof-dev

Sorry, I didn't understand that. Which PR and which merge are you talking about?

Nevertheless I checked out the sof kernel and applied kconfig-sof-default.sh. I am compiling right now ;-)

**Comment by prof-dr-pepper** [comment]
Compilation finished, I installed the kernel and ran alsa-info again: http://alsa-project.org/db/?f=0043c8bc949293f9717707d2c32ec19ee506b525

**Comment by plbossart** [comment]
@prof-dr-pepper I was talking about a future PR, when I have the information needed to create it.

Looks like your build does not use SOF, so something is missing in your config somehow. only HDMI is reported in your results. You should have the module snd-sof-pci-intel-tgl loaded.

here's my local config (a bit behind sof-dev but that should be fine with make olddefconfig):

[config.txt](https://github.com/thesofproject/linux/files/7180599/config.txt)

**Comment by prof-dr-pepper** [comment]
Ok, I tried again with your config. I hope this worked now. Here is the result: http://alsa-project.org/db/?f=8ec1e2d02b170f25c9068f45816d043fb80aa361

**Comment by plbossart** [comment]
> Ok, I tried again with your config. I hope this worked now. Here is the result: http://alsa-project.org/db/?f=8ec1e2d02b170f25c9068f45816d043fb80aa361

something's not going well, still HDMI card. 

please add this to /etc/modprobe.d/alsa-base.conf

blacklist snd-hda-intel

also add if you want to see more, add this file 
[sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/7180981/sof-dyndbg.conf.txt)
to /etc/modprobe.d/sof-dyndbg.conf

and reboot

then check if you see any signs of SOF driver being loaded.

**Comment by prof-dr-pepper** [comment]
I tried using your instructions. Seems that SOF driver is loading now. See: http://alsa-project.org/db/?f=f68c2d585b47fc357bed54b4a897b9095f5c86c7
Here in Germany it's getting late, so I will go to bed now ;-) Let me know if I can help any further. I will try to do so tomorrow or on the weekend.

**Comment by plbossart** [comment]
@prof-dr-pepper can you please try PR #3162? You will need to also do this on your device:

````
sudo cp /lib/firmware/intel/sof-tplg/sof-tgl-rt711-rt1308-mono-rt715.tplg lib/firmware/intel/sof-tplg/sof-tgl-rt711-l0-rt1316-l1-mono-rt714-l3.tplg
````

This command is just to make sure we use a compatible SOF topology file, since we don't have a configuration for this.

**Comment by prof-dr-pepper** [comment]
I compiled your branch, copied the file and tested the kernel. Obviously that didn't do the trick yet, sorry :-/
Here is the result: http://alsa-project.org/db/?f=cf477d37b84bd277bedf2900c2d405e47c8c4140

**Comment by plbossart** [comment]
@prof-dr-pepper well there's progress to some extent, the error happens when trying to load the topology.

````
[    5.438785] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx error for 0x30100000 (msg/reply size: 48/20): -22
[    5.438798] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to load widget PIPELINE.8.HDA2.OUT
[    5.439043] sof-audio-pci-intel-tgl 0000:00:1f.3: error: tplg component load failed -22
[    5.439056] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to load DSP topology -22
[    5.439064] sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at snd_soc_component_probe on 0000:00:1f.3: -22
````

This could be a bug in our recent kernel changes, maybe related to dynamic pipelines. @ranj063 this reminds me of problems we've seen before...

@prof-dr-pepper can you upload the full dmesg log, not the alsa-info results, that'll provide us with more information.

**Comment by plbossart** [comment]
@prof-dr-pepper can you try with this PR https://github.com/thesofproject/linux/pull/3165 , it has fewer of our bleeding edge patches

**Comment by prof-dr-pepper** [comment]
Ok, I will try tomorrow (Saturday) because I am not at home today. Thanks so far. Let me know if you would like to debug live on the system or anything like that. I don't exactly know how we can set this up but maybe you have an idea (maybe kernel debug from another linux system? ...never done this before but should be possible I guess).

**Comment by prof-dr-pepper** [comment]
This is the dmesg with PR #3162 [dmesg.txt](https://github.com/thesofproject/linux/files/7189695/dmesg.txt)

**Comment by prof-dr-pepper** [comment]
This is the dmesg with PR #3162 [dmesg.txt](https://github.com/thesofproject/linux/files/7189727/dmesg.txt)
(sorry, the one I sent earlier was the previous test)

**Comment by prof-dr-pepper** [comment]
Wohooo! You did it! It's working with PR #3165!

Sound on the internal speakers -> works
Plugging in external speakers -> works
Microphone -> works
CPU ->Load is near 0

I'm impressed and super grateful!

Nevertheless I still created the logs for you:
[alsa-info.txt](https://github.com/thesofproject/linux/files/7189765/alsa-info.txt)
[dmesg.txt](https://github.com/thesofproject/linux/files/7189766/dmesg.txt)

So what's happening now? Will this find it's way into the official linux and ubuntu linux kernel somehow and how long could this take? ...or can I get a kernel patch for that so I can apply it to the current latest official linux or even ubuntu linux kernel? (sorry for my excitement... I just didn't expect this to be getting fixed so soon)

**Comment by plbossart** [comment]
@prof-dr-pepper can you double-check if the jack detection works?

I'll double-check with my contacts at Dell that this patch is consistent with the hardware, and send it upstream. In general those quirks are added without delay, so ti's still possible that they land in 5.15, and usually they are back-ported to stable kernels or distro kernels. That's the 'downstream' part on which we don't have much control.

I still have to reproduce the problem where one topology load fails, this could be a backwards compatibility issue of our latest topic/sof-dev kernel with the SOF release 1.8.

**Comment by prof-dr-pepper** [comment]
I already tested that. I played music through the laptop speakers, then I plugged in the external speakers and sound switched over to the external speakers. After that I unplugged the external speakers and sound switched back to the laptop speakers. :-)

**Comment by plbossart** [comment]
I meant testing with a headset with a mic, if you have one, and checking that the capture works on the headset mic. That detection is often problematic.

**Comment by prof-dr-pepper** [comment]
Oh, ok. I could use my smartphone headset for that. I am still at work. I will test that later today :)

**Comment by plbossart** [comment]
@prof-dr-pepper I wasn't able to reproduce the issue you mentioned with PR #3162, so I could use your help to debug further

a) force the SOF trace
Please add this to /etc/modprobe.d/alsa-base.conf
options snd_sof sof_debug=1

b) install 'sof-logger'
If 'sof-logger' is not installed on your system, install it to e.g. /usr/local/bin from https://github.com/thesofproject/sof-bin/tree/main/v1.8.x/tools-v1.8

c) make sure your distro provides /lib/firmware/intel/sof/sof-tgl.ldc (in addition to sof-tgl.ri)
If not, please save /lib/firmware/intel/sof and /lib/firmware/intel/sof-tplg and clone https://github.com/thesofproject/sof-bin, then do 
"sudo ./install.sh v1.8.x/v1.8"

c) re-install PR #3162 (I updated to our latest tip)

d) reboot
attach full dmesg log (not the short version from alsa-info)

extract the firmware trace with
sudo sof-logger -t -l /usr/lib/firmware/intel/sof-v1.8/sof-tgl.ldc | tee sof-log.txt

and attach sof-log.txt

Long sequence but that's just to enable trace capabilities needed to figure out what led to an IPC error.

**Comment by prof-dr-pepper** [comment]
Testing with the microphone of the headset everything seems as I would expect. I tested with "PulseAudio Volume Control". Plug and unplug is detected for the speakers AND for the microphone. As soon as I plug in the headset the microphone of the headset becomes the default audio input and if I tap on the microphone of the headset I can undoubtedly distinguish between sound coming from the headset microphone and the laptop microphone. Unplugging the headset reverses everything and still everything works as expected. Anything else I should test with #3165?

**Comment by plbossart** [comment]
Thanks @prof-dr-pepper, that's good. What we'd need to focus on is PR #3162 since we could have introduced a regression in our latest work and we'd want to avoid breaking your solution with a future kernel update. I wasn't able to reproduce the problem you had with PR #3162, but it's a slightly different device so maybe there's something I missed. Thanks!

**Comment by prof-dr-pepper** [comment]
I'm already on it ;-)
Little update already: *.ldc files were missing

**Comment by prof-dr-pepper** [comment]
With #3162 and the steps that you described above sound seems to work, too. However there is one thing that concerns me. With #3165 I occasionally experienced a temporary freeze of the system which resolved after 10-20 seconds. With #3162 I now experienced a permanent freeze of the system. I don't know if this is caused by your changes in the driver.

**Comment by prof-dr-pepper** [comment]
1. The temporary or permanent freeze does not seem to occur all the time (hard to reproduce... maybe after 50% of the linux reboots). But keep in mind: I haven't used this new laptop regularly yet because I only got it a few days ago. So maybe this happens with the stock ubuntu kernel, too.
2. Furthermore after updating the sof firmware the automatic switch between builtin microphone and headset microphone does not work anymore (for the speakers the automatic switch works) -> neither with #3162 nor with #3165.

**Comment by prof-dr-pepper** [comment]
Here are the logs for the new test with #3162

[alsa-info.txt](https://github.com/thesofproject/linux/files/7198471/alsa-info.txt)
[dmesg.txt](https://github.com/thesofproject/linux/files/7198473/dmesg.txt)
[sof-log.txt](https://github.com/thesofproject/linux/files/7198474/sof-log.txt)

**Comment by plbossart** [comment]
Thanks for testing @prof-dr-pepper.

1) I don't see any IPC errors as in the previous test, do you mind going back to last week's version (commit 8bfbb3d2690c16932570c3e3dba3efbb4f285b74) and recheck

2) this is surprising that we switch doesn't work, in theory this is the same files. Do you mind re-trying as well and checking if the firmware and topology file you use are actually the same between your distribution and the ones from sof-bin?

**Comment by prof-dr-pepper** [comment]
Sure. However it's getting late here, so I will have to do this tomorrow, sorry. :-/

**Comment by prof-dr-pepper** [comment]
This is commit 8bfbb3d with the firmware from sof-bin:
[alsa-info.txt](https://github.com/thesofproject/linux/files/7205980/alsa-info.txt)
[dmesg.txt](https://github.com/thesofproject/linux/files/7205981/dmesg.txt)
[sof-log.txt](https://github.com/thesofproject/linux/files/7205982/sof-log.txt)

Speakers and microphones work but automatic switch between builtin microphone and headset microphone does not work (for the speakers the automatic switch works).

**Comment by prof-dr-pepper** [comment]
Here are the file hashes of the sof firmware (/usr/lib/firmware/intel/sof/) and topology (/usr/lib/firmware/intel/sof-tplg/) files:

distribution:
[dist_sof-tplg.txt](https://github.com/thesofproject/linux/files/7205996/dist_sof-tplg.txt)
[dist_sof.txt](https://github.com/thesofproject/linux/files/7205997/dist_sof.txt)

sof-bin:
[sof-bin_sof-tplg.txt](https://github.com/thesofproject/linux/files/7205998/sof-bin_sof-tplg.txt)
[sof-bin_sof.txt](https://github.com/thesofproject/linux/files/7205999/sof-bin_sof.txt)

**Comment by plbossart** [comment]
Thanks, looks like your distribution did not use SOF release 1.8 then? Can you check with package names/versions?

**Comment by prof-dr-pepper** [comment]
Can I simply revert to the original sof firmware and topology files from the distribution by renaming the directories /usr/lib/firmware/intel/sof/ and topology /usr/lib/firmware/intel/sof-tplg/ ?

I am asking because I would like to test again with #3165 and check again with automatic switch of the mic.

**Comment by plbossart** [comment]
Sure. Let's go back to a stable state for you, then check what happened with #3162 later.

**Comment by prof-dr-pepper** [comment]
Strange... seems like the Ubuntu package firmware-sof-signed (see file list for my distribution here: https://packages.ubuntu.com/hirsute/all/firmware-sof-signed/filelist) which would contain the files is not installed on my system but neverthelesse the sof firmware files and sof tplg files were installed with the distribution. In the original /usr/lib/firmware/intel/sof directory from the distribution there are 3 subdirectories: v1.4.2, v1.6 and v1.6.1 -> So indeed it seems that the distribution comes with an older firmware

**Comment by prof-dr-pepper** [comment]
I renamed the folders back so that I should have the original files from the distribution again and I booted the exact kernel from #3165. Speakers and microphones work fine but still automatic switch of the microphone doesn't work (for the speakers the automatic switch works). This is surprising because in my previous test I remember this to have worked. So either something is different with my current system or I made a mistake during my earlier test. :-/

**Comment by prof-dr-pepper** [comment]
Let me know if I can try anything else for you. I will go offline for today because again here it's getting late now. However as far as I can see I may have more time for testing tomorrow. :)

**Comment by plbossart** [comment]
the switch on microphone doesn't always work, I've seen similar issues in the past where you have to manually select the mic you want. I think Ubuntu are just behind upstream big time.

What I am really interested is the IPC error you had with #3162. If you indeed have the 1.6.1 SOF release, you may be able to use sof-logger with https://github.com/thesofproject/sof-bin/blob/stable-v1.6.1/lib/firmware/intel/sof/v1.6.1/sof-tgl-v1.6.1.ldc

**Comment by prof-dr-pepper** [comment]
Where should I put the sof-tgl-v1.6.1.ldc file? (the directory /lib/firmware/intel/sof/v1.6.1 does not exist)

**Comment by plbossart** [comment]
> Where should I put the sof-tgl-v1.6.1.ldc file? (the directory /lib/firmware/intel/sof/v1.6.1 does not exist)

doesn't really matter, you can put it in your home directory, as long as you use the right path with sof-logger

sudo sof-logger -t -l ~/sof-tgl.ldc | tee sof-log.txt

**Comment by prof-dr-pepper** [comment]
sorry... I'm embarrassed for that... I could have figured ;-)

**Comment by prof-dr-pepper** [comment]
Sorry, that didn't work: error: src hash value from version file (0xb2280b69) differ from src hash version saved in dictionary (0x459cfb8c).

**Comment by prof-dr-pepper** [comment]
However I found the ubuntu package that contains the sof firmware and topology files that are on my system: https://packages.ubuntu.com/hirsute-updates/linux-firmware

**Comment by prof-dr-pepper** [comment]
...i also tried with the embedded /lib/firmware/intel/sof/v1.6/sof-tgl-v1.6.ldc file but that also didn't match

**Comment by prof-dr-pepper** [comment]
Currently I am out of ideas... except to offer you again to have a screen sharing session on my laptop so that you can try for yourself

**Comment by plbossart** [comment]
check in your dmesg log, there should be a line that says which version you are really using.
something like

````
sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware info: version 1:8:0-9e7a8
````

**Comment by prof-dr-pepper** [comment]
It says:
[    5.544219] sof-audio-pci-intel-tgl 0000:00:1f.3: Firmware info: version 1:6:0-18fab

So I tried this file with sof-logger:
https://github.com/thesofproject/sof-bin/blob/stable-v1.6/lib/firmware/intel/sof/v1.6/sof-tgl-v1.6.ldc

But this still resulted in the following error:
error: src hash value from version file (0xb2280b69) differ from src hash version saved in dictionary (0x9de615dc).

Then I also tried this (with the 1.6 file that is shipped with my distro):
sudo sof-logger -t -l /lib/firmware/intel/sof/v1.6/sof-tgl-v1.6.ldc | tee sof-log.txt

But still no success:
error: src hash value from version file (0xb2280b69) differ from src hash version saved in dictionary (0x9de615dc).

**Comment by prof-dr-pepper** [comment]
You know what, I came up with an idea. We had problems with v1.6 firmware from the ubuntu distro running with PR #3162 and we were not able to get the sof-logger running with the firmware from the distro. So I took the v1.6 firmware from the sof-bin repo and tested with it instead (hoping that it would behave similarly). This way I was able to use the sof-logger and hopefully my logfiles are helpful for you. :-)

[alsa-info.txt](https://github.com/thesofproject/linux/files/7230297/alsa-info.txt)
[dmesg.txt](https://github.com/thesofproject/linux/files/7230298/dmesg.txt)
[sof-log.txt](https://github.com/thesofproject/linux/files/7230299/sof-log.txt)

**Comment by plbossart** [comment]
@prof-dr-pepper thanks, that's good but you missed the trick with the topology file, which leads to an error in the dmesg log.
````
error: tplg request firmware intel/sof-tplg/sof-tgl-rt711-l0-rt1316-l1-mono-rt714-l3.tplg failed
````

Please use this command to fix the broken reference:

````
sudo cp /lib/firmware/intel/sof-tplg/sof-tgl-rt711-rt1308-mono-rt715.tplg lib/firmware/intel/sof-tplg/sof-tgl-rt711-l0-rt1316-l1-mono-rt714-l3.tplg
````

**Comment by prof-dr-pepper** [comment]
Doh! ;-) Ok, next try...
[alsa-info.txt](https://github.com/thesofproject/linux/files/7237400/alsa-info.txt)
[dmesg.txt](https://github.com/thesofproject/linux/files/7237401/dmesg.txt)
[sof-log.txt](https://github.com/thesofproject/linux/files/7237402/sof-log.txt)

**Comment by plbossart** [comment]
ok, that's good @prof-dr-pepper. 

I am not sure what happens here. With PR #3165 and the ASoC upstream kernel we have no problems. With PR #3162 which uses a more recent kernel with SOF changes for dynamic pipelines, we have an IPC error

````
[    5.509508] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx: 0x30100000: GLB_TPLG_MSG: PIPE_NEW
[    5.509575] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx error for 0x30100000 (msg/reply size: 48/20): -22
[    5.509600] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to load widget PIPELINE.8.HDA2.OUT
[    5.509923] sof-audio-pci-intel-tgl 0000:00:1f.3: error: tplg component load failed -22
[    5.509943] sof-audio-pci-intel-tgl 0000:00:1f.3: error: failed to load DSP topology -22
````

````
[      139522.390289] (      139227.234375) c0 ipc                           src/ipc/ipc.c:461  ERROR ipc_pipeline_new(): cannot find the scheduling component, pipe_desc->sched_id = 42
[      139535.463205] (          13.072916) c0 ipc                       src/ipc/handler.c:1201 ERROR ipc: pipe 8 creation failed -22
````

@ranj063 does this ring a bell? Wondering if there's an assumption in topology files that we missed?

**Comment by prof-dr-pepper** [comment]
You're welcome. Let me know if there is anything more I can do to help. :)

**Comment by ranj063** [comment]
> @ranj063 does this ring a bell? Wondering if there's an assumption in topology files that we missed?

@plbossart if I look at the dmesg, it looks like we've got a newer kernel with an older firmware. Basically, it seems like the patch to change the list_for_each_entry_reverse() in sof_set_up_pipelines() is incompatible with older firmware. So if you revert this, it might just work.
@prof-dr-pepper would you be able to revert this with PR 3162 and give it a try?
commit 3148a6d4815a3901a1dbedb43268688566789e85
Author: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
Date:   Tue Aug 17 16:24:33 2021 -0500

    ASoC: SOF: sof-audio: don't use list_for_each_entry_reverse()

**Comment by prof-dr-pepper** [comment]
Sure... working on it ;-)

**Comment by prof-dr-pepper** [comment]
I checked out #3162 (plbossart:sdw/fix-dell-sku-0A45), reverted the one line back to list_for_each_entry_reverse and tested again. However audio did not work:
[alsa-info.txt](https://github.com/thesofproject/linux/files/7238010/alsa-info.txt)
[dmesg.txt](https://github.com/thesofproject/linux/files/7238011/dmesg.txt)
[sof-log.txt](https://github.com/thesofproject/linux/files/7238012/sof-log.txt)

**Comment by ranj063** [comment]
What is the ABI version for the firmware you are testing @prof-dr-pepper ? I'm trying to understand why the pipeline widget gets loaded first even with the revert

**Comment by prof-dr-pepper** [comment]
Sorry, I don't know. However the firmware I am using is this one: https://github.com/thesofproject/sof-bin/tree/stable-v1.6
I am using Ubuntu 21.04 which comes with v1.6 of the firmware but I was not able to use sof-logger with the firmware that came with the distro. So I used the one from https://github.com/thesofproject/sof-bin with the same version (hoping that it would behave similarly).

**Comment by ranj063** [comment]
OK 1.6 is older thsn 3.19 I think. In this case, the the pipeline widget should always be set up after the rest of the widgets. So I don;t understand why we're seeing this error in the first place. @plbossart any ideas?

**Comment by plbossart** [comment]
I don't have a clue @ranj063, I'll try with the 1.6 firmware/topology on my CML device and see what happens. We've tested compatibility with 1.8 but not earlier versions.

**Comment by plbossart** [comment]
I can reproduce the issue with topic/sof-dev and firmware/topology 1.6.1

````
[    3.137146] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx: 0x30100000: GLB_TPLG_MSG: PIPE_NEW
[    3.137272] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx error for 0x30100000 (msg/reply size: 48/20): -22
[    3.137315] sof-audio-pci-intel-cnl 0000:00:1f.3: error: failed to load widget PIPELINE.8.HDA2.OUT
[    3.137825] sof-audio-pci-intel-cnl 0000:00:1f.3: error: tplg component load failed -22
[    3.137870] sof-audio-pci-intel-cnl 0000:00:1f.3: error: failed to load DSP topology -22
[    3.137928] sof-audio-pci-intel-cnl 0000:00:1f.3: ASoC: error at snd_soc_component_probe on 0000:00:1f.3: -22
[    3.138049] sof_sdw sof_sdw: ASoC: failed to instantiate card -22
[    3.139914] Failed to get private value: 755800 => 0000 ret=-16
[    3.140504] sof_sdw sof_sdw: snd_soc_register_card failed -22
[    3.140531] sof_sdw: probe of sof_sdw failed with error -22
````

Clearly something we are missing. Thanks a lot @prof-dr-pepper for reporting this, that gives us an opportunity to fix things before polluting the upstream kernel!

**Comment by plbossart** [comment]
The problem happens with firmware 1.6 and 1.7, and goes away with firmware 1.8, where there's no issue

````
[    3.001303] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx: 0x30100000: GLB_TPLG_MSG: PIPE_NEW
[    3.001428] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx succeeded: 0x30100000: GLB_TPLG_MSG: PIPE_NEW
[    3.001432] sof-audio-pci-intel-cnl 0000:00:1f.3: widget PIPELINE.8.HDA2.OUT setup complete
[    3.001438] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx: 0x30010000: GLB_TPLG_MSG: COMP_NEW
[    3.001669] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx succeeded: 0x30010000: GLB_TPLG_MSG: COMP_NEW
[    3.001676] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.001819] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx succeeded: 0x80010000: GLB_DAI_MSG: CONFIG
[    3.001820] sof-audio-pci-intel-cnl 0000:00:1f.3: widget HDA2.OUT setup complete
````

The topology hasn't changed, so it's really something between new kernel and old firmware that's incompatible. Meh.

**Comment by prof-dr-pepper** [comment]
Hey, I am happy that you help me with my driver issues and I am glad that you also benefit from my findings. Let me know if I can be of more help for you. 

Now I have to face the next/bigger challenge: The webcam on this Dell Latitude 7320 Detachable is not working either (see hardware info from my first post). Unfortunately I guess this is not only a bug but there simply is no driver for it yet. I know that this is not your code. However do you guys have any idea which team/project might be responsible for this webcam?
I am quite experienced with cameras myself (as a coder), so I might actually be able to write or assist writing such a driver (or maybe modify an existing one).

**Comment by plbossart** [comment]
@ranj063 I can reproduce this compatibility issue with firmware < 1.8 on UpExtreme, same IPC error on PIPE_NEW

````
ipc tx: 0x30100000: GLB_TPLG_MSG: PIPE_NEW
[    5.466181] sof-audio-pci-intel-cnl 0000:00:1f.3: ipc tx error for 0x30100000 (msg/reply size: 48/20): -22
[    5.466208] sof-audio-pci-intel-cnl 0000:00:1f.3: error: failed to load widget PIPELINE.11.DMIC1.IN
[    5.466565] sof-audio-pci-intel-cnl 0000:00:1f.3: error: tplg component load failed -22
[    5.466592] sof-audio-pci-intel-cnl 0000:00:1f.3: error: failed to load DSP topology -22
[    5.466607] sof-audio-pci-intel-cnl 0000:00:1f.3: ASoC: error at snd_soc_component_probe on 0000:00:1f.3: -22
[    5.466677] skl_hda_dsp_generic skl_hda_dsp_generic: ASoC: failed to instantiate card -22
````

So it's not a problem related to SoundWire or HDaudio but a more fundamental issue.

**Comment by plbossart** [comment]
I moved the IPC error discussion to a different issues https://github.com/thesofproject/linux/issues/3181.

**Comment by plbossart** [comment]
> Hey, I am happy that you help me with my driver issues and I am glad that you also benefit from my findings. Let me know if I can be of more help for you.

Thanks! we'll progress on our own to identify the regression with older firmware.

> Now I have to face the next/bigger challenge: The webcam on this Dell Latitude 7320 Detachable is not working either (see hardware info from my first post). Unfortunately I guess this is not only a bug but there simply is no driver for it yet. I know that this is not your code. However do you guys have any idea which team/project might be responsible for this webcam?
> I am quite experienced with cameras myself (as a coder), so I might actually be able to write or assist writing such a driver (or maybe modify an existing one).

I don't have a clue but @kv2019i might know ?

**Comment by ranj063** [comment]
@plbossart @prof-dr-pepper I have posted a fix for this in https://github.com/thesofproject/linux/pull/3182. So now #3162 + #3182 should work.

**Comment by kv2019i** [comment]
@prof-dr-pepper wrote:
> Now I have to face the next/bigger challenge: The webcam on this Dell Latitude 7320 Detachable is not working either (see 
> hardware info from my first post). Unfortunately I guess this is not only a bug but there simply is no driver for it yet. I know 
> that this is not your code. However do you guys have any idea which team/project might be responsible for this webcam?

Based on hardware info this seems to be a variant of the 7320 with a MIPI connected camera (versus a USB). I know some Chromebooks are using this hardware and there is a driver for IPU6 in https://chromium.googlesource.com/chromiumos/third_party/kernel/+/HEAD/drivers/media/pci/intel/ . This driver is not in upstream kernel yet though, and you do need non-trivial amount of logic in user-space to have a fully working camera, so some support from Ubuntu is needed as well. You could ask for example Sakari Ailus (one of V4L2 kernel maintainers so email in MAINTAINERS file) on the status.

**Comment by prof-dr-pepper** [comment]
I don't know how much the hardware of the normal Latitude 7320 and the Latitude 7320 Detachable are identical but you are correct that the cameras are connected via the MIPI interface (OV5678 sensor for front camera and OV8856 sensor for rear camera).
Thanks for your help. :)

**Comment by prof-dr-pepper** [comment]
@ranj063 @plbossart I can confirm that #3162 + #3182 works. Thank you for your help so that my laptop will have audio in the future. :-)

Here's the proof:
[alsa-info.txt](https://github.com/thesofproject/linux/files/7246040/alsa-info.txt)
[dmesg.txt](https://github.com/thesofproject/linux/files/7246041/dmesg.txt)
[sof-log.txt](https://github.com/thesofproject/linux/files/7246042/sof-log.txt)

It also works with the sof firmware 1.6 that was shipped with my Ubuntu distro and also with sof firmware 1.8 from sof-bin repo. Good job!

**Comment by nos1609** [comment]
Any way to get it woring on Fedora 35?

**Comment by prof-dr-pepper** [comment]
> Any way to get it woring on Fedora 35?

If you wait for the fixes in PR #3162 and #3182 to be merged then I see no reason why you should not be able to compile and use the kernel here to run it on Fedora 35, too. If you don't want to wait you could apply these changes manually (which is what I did).

**Comment by thekswenson** [comment]
> I don't know how much the hardware of the normal Latitude 7320 and the Latitude 7320 Detachable are identical but you are correct that the cameras are connected via the MIPI interface (OV5678 sensor for front camera and OV8856 sensor for rear camera). Thanks for your help. :)

@prof-dr-pepper  I would love to keep up to date with any progress on the webcam front. Could you please direct me towards any bug reports or communications related to the webcam?

**Comment by prof-dr-pepper** [comment]
I didn't investigate any further yet regarding the webcam. I also think that this forum is not the best place for the webcam issue because the sof project is for sound stuff only. Sorry for the sad news

**Comment by thekswenson** [comment]
@prof-dr-pepper   Could you please post the steps you used to get an Ubuntu kernel using these changes?  

I've been trying to use the instructions at https://wiki.ubuntu.com/KernelTeam/GitKernelBuild to clone this topic/sof-dev kernel and configure it for Ubuntu, but have been running into some difficulties.

**Comment by prof-dr-pepper** [comment]
I guess today a new Ubuntu version will be released, so I will install it anyway in the next days (including the sof kernel). I will write down the steps to install the kernel properly and post them here. If you are curious already you can read the history of this forum thread. It should contain all information your need.

Regards, Marcus

**Comment by thekswenson** [comment]
> I guess today a new Ubuntu version will be released, so I will install it anyway in the next days (including the sof kernel). 

I didn't realize that the changes would make it into the Ubuntu kernel so quickly.  That would be great! If that is the case, I won't need to bother with the kernel compilation.

> I will write down the steps to install the kernel properly and post them here. If you are curious already you can read the history of this forum thread. It should contain all information your need.

I've read the thread history a few times while trying to compile the kernel. I'm not using the minimal config that the SOF project gives, since I want all of the functionality of the Ubuntu kernel (I don't want to wade through all the options and hand pick those that I know I will need). The steps I've followed are:

  0. Copy the firmware link according to this thread.
  1. Clone this kernel (the most recent version already has the two important commits merged into it).
  2. Copy the .config file from the Impish (21.10) source 
  3. run `make oldconfig`, making sure to answer 'm' to the SOF questions
  4. run `make -j $(getconf _NPROCESSORS_ONLN) deb-pkg LOCALVERSION=-soffixes`

The kernel compiles but I get some errors from the dpkg-source step.

**Comment by prof-dr-pepper** [comment]
Sorry for the misunderstanding. The upcoming Ubuntu release will not contain the bugfix yet.

**Comment by prof-dr-pepper** [comment]
> > I guess today a new Ubuntu version will be released, so I will install it anyway in the next days (including the sof kernel).
> 
> I didn't realize that the changes would make it into the Ubuntu kernel so quickly. That would be great! If that is the case, I won't need to bother with the kernel compilation.
> 
> > I will write down the steps to install the kernel properly and post them here. If you are curious already you can read the history of this forum thread. It should contain all information your need.
> 
> I've read the thread history a few times while trying to compile the kernel. I'm not using the minimal config that the SOF project gives, since I want all of the functionality of the Ubuntu kernel (I don't want to wade through all the options and hand pick those that I know I will need). The steps I've followed are:
> 
>     1. Copy the firmware link according to this thread.
> 
>     2. Clone this kernel (the most recent version already has the two important commits merged into it).
> 
>     3. Copy the .config file from the Impish (21.10) source
> 
>     4. run `make oldconfig`, making sure to answer 'm' to the SOF questions
> 
>     5. run `make -j $(getconf _NPROCESSORS_ONLN) deb-pkg LOCALVERSION=-soffixes`
> 
> 
> The kernel compiles but I get some errors from the dpkg-source step.

I migrated this fix to the official latest stable Linux Kernel (-> v5.14) and create a little script for it including the patch for the kernel (tested with Kubuntu 21.10). You can find it here:
[build.tar.gz](https://github.com/thesofproject/linux/files/7367169/build.tar.gz)

On my Dell Latitude 7320 Detachable it works fine and the graphics driver works much better with this kernel, too. ;-)

Regards,
Marcus

**Comment by thekswenson** [comment]
> I migrated this fix to the official latest stable Linux Kernel (-> v5.14) and create a little script for it including the patch for the kernel (tested with Kubuntu 21.10). You can find it here: [build.tar.gz](https://github.com/thesofproject/linux/files/7367169/build.tar.gz)
> 

Thanks @prof-dr-pepper !

I have sound on this end :).  Appreciate the help. I didn't realize that it would be easier to grab the generic kernel, rather than the one from Ubuntu.


> On my Dell Latitude 7320 Detachable it works fine and the graphics driver works much better with this kernel, too. ;-)

I wasn't having any graphics problems until I updated to the 5.13 kernel (when updating to Ubuntu 21.10). The problems were related to an intel graphics power saving thing called Panel Self Refresh (PSR). I had to disable PSR when booting the kernel (see https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1947326). I confirmed that PSR was actually working with the 5.11 kernel but it appears to no long be working with newer kernels, including the 5.14 that we just compiled :(. I suppose I'll have to just boot the 5.11 if I'm going to be without a power source for an extended period.


I have a few questions about the sound:

  1. Are you having balance issues with your speakers?  My left speaker is clearly _much_ quieter than my right one (try to hear Cannonball Adderly in All Blues, or go to sound settings and click the "front left" button). With headphones plugged in the balance is right though.
  2. Does your machine switch over to the headphones when plugged in?  I can confirm with pavucontrol that both the headphones and microphone on my headset are detected when I plug it in.  I have to manually select them, however, to activate them. Reading the thread, it looks like it switched automatically for you.
  3. What happens when you activate s2idle while sound is playing?  All hell breaks lose on my end.

**Comment by thekswenson** [comment]
>     1. Are you having balance issues with your speakers?  My left speaker is clearly _much_ quieter than my right one (try to hear Cannonball Adderly in All Blues, or go to sound settings and click the "front left" button). With headphones plugged in the balance is right though.
> 

It turns out this point is just a bad surprise. I booted into Windows (for the second time since I had the machine) and it turns out I have the same problem in Windows! Looks like there may be a manufacturing defect :(.

>     2. Does your machine switch over to the headphones when plugged in?  I can confirm with pavucontrol that both the headphones and microphone on my headset are detected when I plug it in.  I have to manually select them, however, to activate them. Reading the thread, it looks like it switched automatically for you.
> 
>     3. What happens when you activate s2idle while sound is playing?  All hell breaks lose on my end.

I have a fourth thing to add to this list.

  4. The speakers works only  part of the time. For example, I just booted into the 5.14 kernel and no sound was coming from the speakers, despite being selected in pavucontrol. So everything _looked_ normal, and the headphones were working (when I switch them manually), yet no sound would come out when selecting "Tiger Lake-LP Smart Sound Technology Audio Controller". I put the thing to sleep with s2idle and upon wake the sound worked!  Do you experience these kinds of instabilities?

**Comment by prof-dr-pepper** [comment]
I haven't tested much yet. But today I once noticed the issue that no sound was coming out of the speakers after reboot. Plugging in the headphones and also microphone worked.
I will have a look at the other things you mentioned as soon as I have enough time. I hope I can do it by the end of the week.

**Comment by prof-dr-pepper** [comment]
I just compiled the current sof kernel and it seems to work much better.
Here is the script and the patch that I used:
[build.tar.gz](https://github.com/thesofproject/linux/files/7374035/build.tar.gz)

**Comment by thekswenson** [comment]
> I just compiled the current sof kernel and it seems to work much better. Here is the script and the patch that I used: [build.tar.gz](https://github.com/thesofproject/linux/files/7374035/build.tar.gz)

Yup. The SOF kernel is much better. It reacts normally when going s2idle, and sound hasn't inexplicably cut out yet. Thanks!

So thing are working as expected at this point, except for automatic switch to headphones. Does your machine switch to headphones automatically when you plug them in?

I'm also curious to know if your left speaker is much quieter than your right. I'm wondering if yours is like this? If so, maybe there is a general flaw, rather than just a manufacturing defect or the like.

**Comment by prof-dr-pepper** [comment]
Yes, the automatic switch for the headphones works and the speakers seem equally loud. Seems to be an issue with your device, sorry.

**Comment by thekswenson** [comment]
Thanks for the feedback!

For the headphones: this seems like a software issue, given that they show up as headphones and I can choose them and then use them normally. Is it the kernel module that controls the automatic switching or is it more of a pulseaudio issue?

**Comment by plbossart** [comment]
The speakers and headphones are handled by different chips, and can operate in parallel if needed. 
You can test this with the two terminals using the following commands.
speaker-test -Dhw:0,0 -c2 -r48000 
speaker-test -Dhw:0,2 -c2 -r48000

There is no mutual exclusion or switch handled at the kernel level.

IIRC there is a PulseAudio module-switch-on-connect or something.

It's been my experience that the headphone switch works fine. 

The headset mic is more complicated, I think in the OEM image you get a screen menu asking if the device is a headphone or headset (with mic), but I never figured out what you need to do to get this behavior in other distros.

**Comment by prof-dr-pepper** [comment]
@plbossart The bugfix that you created for this issue here requires a toplology file named "sof-tgl-rt711-l0-rt1316-l1-mono-rt714-l3.tplg". This file does neither seem to be shipped with any Linux distribution nor does it seem come with the firmware from https://github.com/thesofproject/sof-bin.
Will the missing file be added some day somewhere?

**Comment by plbossart** [comment]
Thanks @prof-dr-pepper for reporting this, this seems to be a release backport issue. I did add support for this file in the tools/topology/topology1/CMakeLists.txt file (SOF commit 425743e772342256141201a74c760551f1953561 "topology1: add missing Dell SKU 0A45 topology")

````
	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-tgl-rt711-l0-rt1316-l1-mono-rt714-l3\;-DPLATFORM=tgl\;-DMONO\;-DUAJ_LINK=0\;-DAMP_1_LINK=1\;-DMIC_LINK=3"
````

but this file is indeed not in the sof-bin release.

@lgirdwood @marc-hb could it be that this topology update was not added to the 1.9.2 branch? I see 10 differences between main and v1.9.2, not sure why the topology changes were left out?

**Comment by marc-hb** [comment]
I see only one commit from @plbossart in https://github.com/thesofproject/sof/commits/stable-v1.9 right now and it's indeed not a topology change.

> not sure why the topology changes were left out?

Simply because no one backported them? I'm not aware of any systematic backporting process, AFAIK what is in stable-v1.9 is only what people explicitly submit to that branch.
